### ISSUE TITLE
hie-core test cases that expect 0 diagnostics always succeed

### ISSUE THREAD

**Issue by hsenag** [digital-asset/daml#2813]
If you edit `compiler/hie-core/test/exe/Main.hs` to make one of the tests
expect no diagnostics, then it hits the base case of `Development.IDE.Test.expectDiagnostics` and doesn't check that there are actually 0 diagnostics.

I think it's because it just hits the `Map.null` base case immediately. Also it would go wrong if there are diagnostics for files not listed in the `expected` list that come at the end of the published diagnostics.

I'm happy to have a go at fixing it myself if someone could confirm my analysis so far is correct.

**Comment by cocreature** [comment]
Your analysis is completely right. `expectDiagnostics` is definitely a rather fragile beast and could use some improvements.

**Comment by hsenag** [comment]
OK, I see part of the implementation difficulty is that diagnostics just arrive on an asynchronous stream from the server.  Can we expect [`waitForDiagnostics`](http://hackage.haskell.org/package/lsp-test-0.6.1.0/docs/Language-Haskell-LSP-Test.html#v:waitForDiagnostics) to do the right thing, so we can collect all the "actuals" up front and then start comparing?

**Comment by cocreature** [comment]
Not quite sure what you mean by “the right thing” here. The main reason why it’s tricky to get this right is that:

1. We only emit diagnostics if they changed. In particular, you will not get any diagnostics after loading a file without any errors and then making a change that still doesn’t produce errors.
2. Diagnostics are stored per version and emitted (potentially) separately per stage. What this means is that when a document is updated, its version changed so initially we reset the diagnostics to being empty. As we walk through the dependency graph, we will update the diagnostics with the diagnostics we encounter at parsing, typechecking, …. Each of those stages will emit diagnostics if they changed. So it can happen that you have a file with a type error, you modify it, the diagnostics are reset to being empty, we publish empty diagnostics, then you get to parsing which still has empty diagnostics, and then we get to typechecking where the diagnostics change again to being non-empty and we publish that.

We’ve added progress notifications in the meantime so what might work is to accumulate all diagnostics until you get the progress end notification and then look at the latest value.

**Comment by hsenag** [comment]
Right, I realised in the meantime that `waitForDiagnostics` just does the same thing as the existing
code, so certainly wouldn't have helped.

Thanks for the hints, I'l take a look at using the progress notifications.

**Comment by hsenag** [comment]
Unfortunately I don't think it's guaranteed that the ProgressDone comes after the last PublishDiagnostics, because they are coming from different threads and the notifications are de-bounced.  Certainly in practice I see some diagnostics coming after the last ProgressDone.

The best idea I have left is a short sleep at the very end of each `testSession` followed by checking for any remaining messages and erroring if they are diagnostics.

But even that is a bit tricky because there isn't a trivial way to just look for outstanding messages and return immediately. [`withTimeout`](http://hackage.haskell.org/package/lsp-test-0.7.0.0/docs/Language-Haskell-LSP-Test.html#v:withTimeout) looked promising but I couldn't see a way to catch the resulting `Timeout` exception within the `Session` monad.

What I do have working is to sleep a bit, then send a dummy request that will result in some kind of response. Once we see the response to the dummy request we finish, but if in the meantime we see any unexpected diagnostics we can report an error.

What do you think? The length of the sleep will need to be tuned to balance detecting problems versus slowing each test down. Time-based tests are always nasty so maybe it's better to just live with the current situation if noone has any better ideas.

**Comment by cocreature** [comment]
I’m not quite sure how the dummy request helps. You can still get diagnostics after you get the response no?

**Comment by hsenag** [comment]
The intended logic is to look for diagnostics that arrive within a defined time window (e.g. 1s),
and then assume no more are coming.

I can't just use `anyMessage` because it blocks indefinitely if there aren't actually any messages
(or rather it hits a timeout and throws an exception, which I can't easily handle).

So instead I sleep for the 1s window, then send a dummy request, then wait for messages. This is on the assumption that any diagnostics that were received during that 1s window will be seen by `anyMessage` before the response to the dummy request.

**Comment by cocreature** [comment]
I’m not too keen on switching from a test that doesn’t rely on timing to one that does but if I understand you correctly you are proposing to switch from the current setup which can succeed when it shouldn’t to one that will fail if the timing works out or behave as it does currently. That does seem like a clear improvement over the status quo.

**Comment by hsenag** [comment]
Yes, that's right, it's a strict improvement in functionality.

However, it will lead to each test taking longer. It looks like 0.5s is about the right timeout to get failures most of the time, which roughly doubles the total running time of the test harness from 4s to 8s on my machine.

Also, tests that can fail non-deterministically often have wider costs. For example it makes it more likely that someone will make a change that breaks the tests without knowing it. Once HEAD is broken you have all sorts of pain for everyone. In the end it might have been better just to risk having some bugs.

Anyway, I'll send a PR so you can get a more concrete understanding of the idea.
